### ISSUE TITLE
Web URL static identifiers (supposedly permalinks) are changing 

### ISSUE THREAD

**Issue by petrs** [crocs-muni/sec-certs#443]
**Describe the bug**
The auto generated ids for sec-certs urls pointing to specific certified product are not always persistent in time (https://sec-certs.org/cc/**6207e27699db6695**/). We do have at least two cases of id that changed in time - for example https://sec-certs.org/cc/6207e27699db6695/ vs. https://sec-certs.org/cc/56ef6c45d7eab7ee/ . The ID is generated from presumably static parameters of certificates, but it changed anyway. If these changes, we are loosing permalink people may have in their papers or watched certificates (email/RSS/Compare). We shall at least detect and add redirect. (and also investigate what is changing)

**To Reproduce**
1. Find some certificate (e.g., Strong Customer Authentication for Apple Pay on Apple Watch with S8 running watchOS 9.4(watchOS 9.4 (build 20T253)) ( ANSSI-CC-2023/60) https://sec-certs.org/cc/56ef6c45d7eab7ee/)
2. Wait some time
3. Check again based on url - if bug demonstrates itself, you will get Not Found
4. Search for certified device name again => observe different url https://sec-certs.org/cc/6207e27699db6695/

**Expected behavior**
Once indexed and inserted to database, the url static string shall stay same. 
If change is anyway detected, notify admins and add automatic redirect to new url. 

**Desktop (please complete the following information):**
- web at sec-certs.org (end of 2023/begin 2024 -> October 2024)

**Additional context**
Known examples
NXP JCOP 4 P71 https://sec-certs.org/cc/2a45531c2dbd1ab8/ -> https://sec-certs.org/cc/f29f88756682e034/
 Strong Customer Authentication for Apple Pay on Apple Watch with S8 running watchOS 9.4(watchOS 9.4 (build 20T253)) ( ANSSI-CC-2023/60) https://sec-certs.org/cc/56ef6c45d7eab7ee/ -> https://sec-certs.org/cc/6207e27699db6695/

**Comment by adamjanovsky** [comment]
This one will be tough to fix though, it's pretty deep in our design and we've repeatedly revisited this with no resolution. 

@petrs If this is important to you, could you please fit it to Thu team meeting agenda? We should discuss possible solutions in person.

**Comment by petrs** [comment]
> This one will be tough to fix though, it's pretty deep in our design and we've repeatedly revisited this with no resolution.
> @petrs If this is important to you, could you please fit it to Thu team meeting agenda? We should discuss possible solutions in person.

Yes, I thing this is important feature to have (permalinks). It breaks email notification functionality as well, so it is real bug. Let's discuss in person

**Comment by J08nY** [comment]
Let me put some notes here that I made on Matrix, just to not lose them.

During normal operation of the site it will never happen that a previously valid certificate pages gives a 404. If a certificate is removed from an update run (or its ID changes, etc., we can not really tell these things apart) the only thing that happens is that a note is added to the certificate update log, (see [here](https://sec-certs.org/cc/8c7ccf48dcbd5398/) for an example, [here](https://sec-certs.org/admin/update/diff/670460332ec22c416676abe9) is the diff).

## What happened to lead to the 404s

Some time ago CC portal changed the way they construct URLs to certification artifacts (PDFs). We were including these links in our primary key that we build the ID out of. Thus one update caused almost the whole database to turn over: "old" certificates were marked as removed and "new" certificates with new IDs took their place. The database was thus sort of duplicated. To clean this up, we decided to make a new fresh run and clear up the database (as well as make our ID construction algorithm agnostic to these sort of URL changes). We also added automated redirects from the old IDs to the new ones.

However, the CC portal also changed some other attributes of some certificates during this change. Since we did the fresh run we lost some certificates here, because the automated redirects were put in based on the new atrtributes (and lead to the wrong "old" IDs).

## How to avoid this in the future?

Do not do fresh runs.

## How to fix this now

Get pre-update dataset, examine which of its IDs are giving 404, quantify and create more redirects manually?

## Bottom line

Even if we do no fresh runs in the future. The isue is not resolved fully.  Even if we had not done the fresh run which has caused the 404s, we would have had a different problem. The page with the old id would list the certificate as removed and there would be a new certificate in the database (and a page) with the new id. This is because to us, these are two different certificates, even though your human eye can tell they are the same. See below for a case study on this.

## Case study

Name and cert ID change. :arrow_right: 
`STMicroelectronics ST33K1M5A and ST33K1M5M B02` to `STMicroelectronics ST33K1M5A and ST33K1M5M B03`
 - Removed: https://sec-certs.org/cc/8c7ccf48dcbd5398/, NSCIB-CC-2300112-01-CR
 - New: https://sec-certs.org/cc/83fc6586c8ac42f5/, NSCIB-CC-2300112-02-CR

Name and cert ID change. :arrow_right: 
`STMicroelectronics ST33K1M5C and ST33K1M5T C01` to `STMicroelectronics ST33K1M5C and ST33K1M5T C02`
 - Removed: https://sec-certs.org/cc/a6f7f2facccfcb2c/, NSCIB-CC-2300056-01-CR
 - New: https://sec-certs.org/cc/34d1675a7a079459/, NSCIB-CC-2300056-02-CR

Name and cert ID stayed. :heavy_equals_sign: Security target URL changed (`...v1.0` to `...v1.1`)
`Junos OS 22.3R1 for ACX5448-M, EX4400-24T, EX4400-24P, EX4400-24MP, EX4400-48T, EX4400-48P, EX4400-48F, EX4400-48MP and QFX5120-YM`
 - Removed: https://sec-certs.org/cc/9246f9e02631f69e/, Certificate Number: 2024/148
 - New: https://sec-certs.org/cc/7ff6ceb530a8d40b/, Certificate Number: 2024/148

**Comment by J08nY** [comment]
Looking at the case studies I think a proper fix is to rethink what a certificate even is. In the cases above, which cases represent the same certificate and which present a new one? For example the first two where part of the cert ID changed are quite weird. See the respective certificates: [old](https://www.commoncriteriaportal.org/nfs/ccpfiles/files/epfiles/NSCIB-CC-2300112-01-CERT.pdf) [new](https://www.commoncriteriaportal.org/nfs/ccpfiles/files/epfiles/NSCIB-CC-2300112-02-CERT.pdf).

Once we understand what a certificate even is, we can think about the changes to our methodology necessary to reflect that. I think it would have to go in the direction of more similarity based metrics and fuzzy stuff rather than hard IDs. Also, the tool is currently quite naive with respect to incremental updates and the web is doing a fresh run each time + some custom update tracking. This should be aligned, because without time information about the previous state of the dataset it is hard to say that some entries are really the same certificate.

**Comment by petrs** [comment]
> Looking at the case studies I think a proper fix is to rethink what a certificate even is. In the cases above, which cases represent the same certificate and which present a new one? For example the first two where part of the cert ID changed are quite weird. See the respective certificates: [old](https://www.commoncriteriaportal.org/nfs/ccpfiles/files/epfiles/NSCIB-CC-2300112-01-CERT.pdf) [new](https://www.commoncriteriaportal.org/nfs/ccpfiles/files/epfiles/NSCIB-CC-2300112-02-CERT.pdf).

I checked these two and I believe these shall be considered two different certificates:
- cert ID is different (although only suffix) NSCIB-2300112-01 vs. NSCIB-2300112-02 
- Name of product is different (ST33K1M5M B02 vs. ST33K1M5M B03)
- Assurance level/package is different (EAL 6 augmented with ALC_FLR.1 vs. EAL6 augmented with ALC_FLR.2)

**Comment by J08nY** [comment]
I made the following change to the website to handle the cases where some of the primary key attributes changed, but not much. @petrs Is this a good-enough resolution for you?

![image](https://github.com/user-attachments/assets/4ee420b2-6d5a-4bf1-b4f4-274fe0d06cd4)

**Comment by J08nY** [comment]
I closed this as stale. It was essentially resolved. If something bad happens again we can reopen and handle it.
